packages/xstate-store/src/index.ts

```diff
@@ -0,0 +1,226 @@
+
+  function receive(event: StoreEvent) {
+    let currentContext = currentSnapshot.context;
+    const assigner = transitions?.[event.type as StoreEvent['type']];
+
+    if (!assigner) {
+      return;
+    }
+
+    if (typeof assigner === 'function') {
+      currentContext = updater
+        ? updater(
+            currentContext,
+            (draftContext) =>
+              (assigner as StoreCompleteAssigner<TContext, StoreEvent>)?.(
+                draftContext,
+                event
+              )
+          )
+        : setter(currentContext, (draftContext) =>
+            Object.assign(
+              {},
+              currentContext,
+              assigner?.(
+                draftContext,
+                event as any // TODO: help me
+              )
+            )
+          );
+    } else {
+      const partialUpdate: Record<string, unknown> = {};
+      for (const key of Object.keys(assigner)) {
+        const propAssignment = assigner[key];
+        partialUpdate[key] =
+          typeof propAssignment === 'function'
+            ? (
+                propAssignment as StorePartialAssigner<
+                  TContext,
+                  StoreEvent,
+                  typeof key
+                >
+              )(currentContext, event)
+            : propAssignment;
+      }
+      currentContext = Object.assign({}, currentContext, partialUpdate);
+    }
+
+    currentSnapshot = { ...currentSnapshot, context: currentContext };
+
+    observers?.forEach((o) => o.next?.(currentSnapshot));
+  }
+
+  const store: Store<TContext, StoreEvent> = {
+    send(event) {
+      receive(event as unknown as StoreEvent);
+    },
+    getSnapshot() {
+      return currentSnapshot;
+    },
+    getInitialSnapshot() {
+      return initialSnapshot;
+    },
+    subscribe(observerOrFn) {
+      const observer = toObserver(observerOrFn);
+      observers ??= new Set();
+      observers.add(observer);
+
+      return {
+        unsubscribe() {
+          return observers?.delete(observer);
+        }
+      };
+    },
+    [symbolObservable](): InteropSubscribable<StoreSnapshot<TContext>> {
+      return this;
+    }
+  };
+
+  return store;
+}
+
+/**
+ * Creates a `Store` with a provided producer (such as Immer's `producer(…)`
+ * A store has its own internal state and can receive events.
+ * 
+ * @example
+  ```ts
+  import { produce } from 'immer';
+
+  const store = createStoreWithProducer(produce, {
+    // initial context
+    { count: 0 },
+    // transitions 
+    {
+      on: {
+        inc: (context, event: { by: number }) => {
+          context.count += event.by;
+        }
+      }
+    }
+  });
+
+  store.getSnapshot(); // { context: { count: 0 } }
+  store.send({ type: 'inc', by: 5 });
+  store.getSnapshot(); // { context: { count: 5 } }
+  ```
+ */
+export function createStoreWithProducer<
+  TProducer extends <TContext>(
+    context: TContext,
+    recipe: (context: TContext) => void
+  ) => TContext,
```

Comment:
does this have to be generic? 🤔 it feels like u could replace this with a simpler:
```ts
  producer: (context: TContex, recipe: (context: TContext) => void) => TContext,
```

Reply:
Ah yes you're right